core/modules.rs

```diff
@@ -680,13 +683,42 @@ impl Stream for RecursiveModuleLoad {
   }
 }
 
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub(crate) enum AssertedModuleType {
+  JavaScriptOrWasm,
+  Json,
+}
+
+// NOTE(bartlomieju): we actually want to have `Into` implementation, instead
+// of `From`, because this operation shouldn't work two-ways, ie. you shouldn't
+// be able to go from `ModuleType` into `AssertedModuleType`. This should
+// probably be removed once WASM module support lands.
+#[allow(clippy::from_over_into)]
+impl Into<AssertedModuleType> for ModuleType {
+  fn into(self) -> AssertedModuleType {
+    match self {
+      ModuleType::JavaScript => AssertedModuleType::JavaScriptOrWasm,
+      ModuleType::Json => AssertedModuleType::Json,
+    }
+  }
+}
+
+impl std::fmt::Display for AssertedModuleType {
+  fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+    match self {
+      Self::JavaScriptOrWasm => write!(f, "JavaScriptOrWasm"),
+      Self::Json => write!(f, "JSON"),
+    }
+  }
+}
+
 /// Describes what is the expected type of module, usually
 /// it's `ModuleType::JavaScript`, but if there were import assertions
 /// it might be `ModuleType::Json`.
```

Comment:
```suggestion
/// Describes a request for a module as parsed from the source code.
/// Usually executable (`JavaScriptOrWasm`) is used, except when an
/// import assertions explicitly constrains an import to JSON, in
/// which case this will have a `AssertedModuleType::Json`.
```